src/contract/event.rs

```diff
@@ -516,12 +540,115 @@ impl<T: Transport, E: ParseLog> AllEventsBuilder<T, E> {
+                //   it.
+                return self.query().await;
+            }
+        };
+        if let Some(deployment_tx) = self.deployment_transaction {
+            start_block = cmp::max(
+                start_block,
+                block_number_from_transaction_hash(self.web3.clone(), deployment_tx).await?,
+            );
+        }
+
+        let end_block = match self.to_block {
+            None | Some(BlockNumber::Latest) | Some(BlockNumber::Pending) => {
+                // NOTE: For latest and pending blocks, we need to set a target
+                //   for the pagination, the last query will use "latest" or
+                //   "paginated" ensuring that all expected events are produced.
+                self.web3.eth().block_number().compat().await?.as_u64()
+            }
+            Some(BlockNumber::Number(value)) => value.as_u64(),
+            Some(BlockNumber::Earliest) => {
+                // NOTE: Query doesn't really make sense, let the node deal with
+                //   it.
+                return self.query().await;
+            }
+        };
+
+        if start_block > end_block {
+            return Ok(Vec::new());
+        }
+
+        let page_size = self.block_page_size.unwrap_or(DEFAULT_BLOCK_PAGE_SIZE);
+        let (web3, filter) = self.prepare();
+
+        let mut events = Vec::new();
+        let mut append_events = |logs: Vec<Log>| -> Result<(), ExecutionError> {
+            events.reserve(logs.len());
+            for log in logs {
+                let event = Event::from_log(log, E::parse_log)?;
+                events.push(event);
+            }
+            Ok(())
+        };
+
+        // NOTE: Query only until the page right before the last one, since it
+        //   is handled a little differently to deal with "latest" and "pending"
+        //   to blocks.
+        let mut current_block = start_block;
+        while current_block + page_size <= end_block {
+            let filter = filter
+                .clone()
+                .from_block(current_block.into())
+                .to_block((current_block + page_size - 1).into())
+                .build();
+            let event_page = web3.eth().logs(filter).compat().await?;
+            append_events(event_page)?;
+
+            current_block += page_size;
+        }
+
+        // NOTE: The last page is handled a bit differently by using the
+        //   `to_block` that was originally specified to the builder and is set
+        //   in the `filter` (from the call to `prepare`). This is done in case
+        //   the to block was "latest" or "pending", where we want to make sure
+        //   that the last call includes blocks that have been added since the
+        //   start of the paginated query.
+        let filter = filter.from_block(current_block.into()).build();
+        let event_page = web3.eth().logs(filter).compat().await?;
+        append_events(event_page)?;
```

Comment:
I see now. Maybe it would be ok to have this logic as a nested if inside the while? 

```rust
// Until we reach the specified end block limit block range to page size.
while current_block <= end_block {
  ...
  if current_block + page_size <= end_block {
    filter.to_block((current_block + page_size - 1).into())
  }
  ...
}
```

Reply:
The issue there is that if `to_block` is a number and not "latest" or "pending" we would do the page again. The condition would have to be a little different.
```
while current_block <= end_block {
  ...
  if current_block + page_size <= end_block && unbound_to_block {
    filter.to_block((current_block + page_size - 1).into())
  }
  ...
}
```

Reply:
Actually we don't need the `bounded_end_block` check! Nice I like this better.